scene/main/viewport.cpp

```diff
@@ -3234,14 +3239,18 @@ void Viewport::_push_unhandled_input_internal(const Ref<InputEvent> &p_event) {
 	}
 
 	if (physics_object_picking && !is_input_handled()) {
-		if (Input::get_singleton()->get_mouse_mode() != Input::MOUSE_MODE_CAPTURED &&
-				(Object::cast_to<InputEventMouse>(*p_event) ||
-						Object::cast_to<InputEventScreenDrag>(*p_event) ||
-						Object::cast_to<InputEventScreenTouch>(*p_event)
+		if (Object::cast_to<InputEventMouse>(*p_event) ||
+				Object::cast_to<InputEventScreenDrag>(*p_event) ||
+				Object::cast_to<InputEventScreenTouch>(*p_event)) {
+			ProjectSettings *ps = ProjectSettings::get_singleton();
+			const bool is_captured = Input::get_singleton()->get_mouse_mode() == Input::MOUSE_MODE_CAPTURED;
 
-								)) {
-			physics_picking_events.push_back(p_event);
-			set_input_as_handled();
+			if (is_captured && ps->has_setting(LEGACY_INPUT_SETTING) && ps->get_setting(LEGACY_INPUT_SETTING).is_one()) {
+				return;
```

Comment:
You can pull this outside the two `if`s above and return early. It avoids another layer of indentation.

Maybe it's also worth having a function that returns the result of this since it is a non-trivial boolean condition and is also used in `Viewport::_process_picking()`. That way you can just call that function. It can be local to this file by being labelled `static` and not being a member of `Viewport`.